src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -245,72 +245,48 @@ void addArg(RefKind refKind, BoundExpression expression)
-                        {
-                            // array[int]
-                            Debug.Assert(input.Type?.IsSZArray() == true);
-                            BoundExpression access = _factory.ArrayAccess(input, makeImplicitIndexArgument(e.Index, e.LengthTemp));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
-
-                    case BoundDagIndexerEvaluation { IndexerSymbol: PropertySymbol indexer } e:
-                        {
-                            // this[int]
-                            BoundExpression access = _factory.Indexer(input, indexer, makeImplicitIndexArgument(e.Index, e.LengthTemp));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
-
-                    case BoundDagIndexerEvaluation { IndexerAccess: BoundIndexerAccess indexerAccess } e:
+                    case BoundDagIndexerEvaluation e:
                         {
                             // this[Index]
-                            BoundExpression access = makeIndexerAccess(e.Syntax, indexerAccess, makeExplicitIndexArgument(e.Index, e.Index < 0));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
+                            // this[int]
+                            // array[Index]
+                            _localRewriter.AddPlaceholderReplacement(e.ReceiverPlaceholder, input);
+                            _localRewriter.AddPlaceholderReplacement<BoundListPatternUnloweredIndexPlaceholder>(e.ArgumentPlaceholder, makeUnloweredIndexArgument(e.Index));
 
-                    case BoundDagSliceEvaluation { IndexerAccess: null, SliceMethod: null } e:
-                        {
-                            // RuntimeHelpers.GetSubArray(T[], Range)
-                            TypeSymbol inputType = input.Type;
-                            Debug.Assert(inputType is { });
-                            Debug.Assert(inputType.IsSZArray());
-                            Debug.Assert(e.StartIndex >= 0 && e.EndIndex <= 0);
+                            var indexerAccess = e.IndexerAccess;
+                            if (indexerAccess is BoundIndexOrRangePatternIndexerAccess implicitAccess)
+                            {
+                                indexerAccess = implicitAccess.WithLengthOrCountAccess(_tempAllocator.GetTemp(e.LengthTemp));
```

Comment:
>_tempAllocator.GetTemp(e.LengthTemp) [](http://example.com/codeflow?start=87&length=36)

Isn't this producing a lowered node? If so, it shouldn't be subjected to lowering again. Since we are handling ```BoundIndexOrRangePatternIndexerAccess``` specially here anyway, perhaps we can extract a shared helper that can be called explicitly from here and from regular visit process. The helper will be instructed if the length access should be lowered or not. The same flag can be used to suppress caching, I think.

Reply:
On the other hand, I think it is fine to keep this as is.